tapdb/universe_perf_test.go

```diff
@@ -0,0 +1,402 @@
+
+// executeSQLStatements executes a list of SQL statements with logging.
+func executeSQLStatements(t *testing.T, db *BaseDB, statements []string,
+	action string) {
+
+	t.Logf("%s indices...", action)
+
+	for _, stmt := range statements {
+		_, err := db.Exec(stmt)
+		if err != nil {
+			t.Fatalf("Failed to %s index: %v\nStatement: %s",
+				action, err, stmt)
+		}
+	}
+}
+
+// createIndices creates all required indices on the database.
+func createIndices(t *testing.T, db *BaseDB) {
+	statements := []string{
+		createUniverseRootIndex,
+		createUniverseLeavesIndex,
+		createMSMTNodesIndex,
+	}
+
+	executeSQLStatements(t, db, statements, "Creating")
+}
+
+// dropIndices removes all custom indices from the database.
+func dropIndices(t *testing.T, db *BaseDB) {
+	statements := []string{
+		`DROP INDEX IF EXISTS idx_universe_roots_composite`,
+		`DROP INDEX IF EXISTS idx_universe_leaves_composite`,
+		`DROP INDEX IF EXISTS idx_mssmt_nodes_composite`,
+	}
+
+	executeSQLStatements(t, db, statements, "Dropping")
+}
+
+// Query definitions for performance testing.
+const (
+	// fetchUniverseRootQuery returns root info for a given namespace.
+	fetchUniverseRootQuery = `
+		SELECT 
+			universe_roots.asset_id, 
+			group_key, 
+			proof_type,
+			mssmt_nodes.hash_key root_hash, 
+			mssmt_nodes.sum root_sum,
+			genesis_assets.asset_tag asset_name
+		FROM universe_roots
+		JOIN mssmt_roots 
+			ON universe_roots.namespace_root = mssmt_roots.namespace
+		JOIN mssmt_nodes 
+			ON mssmt_nodes.hash_key = mssmt_roots.root_hash 
+			AND mssmt_nodes.namespace = mssmt_roots.namespace
+		JOIN genesis_assets
+			ON genesis_assets.asset_id = universe_roots.asset_id
+		WHERE mssmt_nodes.namespace = $1`
+
+	// queryUniverseLeavesQuery gets leaf info for a namespace.
+	queryUniverseLeavesQuery = `
+		SELECT 
+			leaves.script_key_bytes, 
+			gen.gen_asset_id, 
+			nodes.value AS genesis_proof, 
+			nodes.sum AS sum_amt, 
+			gen.asset_id
+		FROM universe_leaves AS leaves
+		JOIN mssmt_nodes AS nodes
+			ON leaves.leaf_node_key = nodes.key 
+			AND leaves.leaf_node_namespace = nodes.namespace
+		JOIN genesis_info_view AS gen
+			ON leaves.asset_genesis_id = gen.gen_asset_id
+		WHERE leaves.leaf_node_namespace = $1 
+			AND (leaves.minting_point = $2 OR $2 IS NULL)
+			AND (leaves.script_key_bytes = $3 OR $3 IS NULL)`
+
+	// queryAssetStatsQuery gets aggregated stats for assets.
+	queryAssetStatsQuery = `
+		WITH asset_supply AS (
+			SELECT 
+				gen.asset_id, 
+				SUM(nodes.sum) AS supply
+			FROM universe_leaves leaves
+			JOIN universe_roots roots
+				ON leaves.universe_root_id = roots.id
+			JOIN mssmt_nodes nodes
+				ON leaves.leaf_node_key = nodes.key 
+				AND leaves.leaf_node_namespace = nodes.namespace
+			JOIN genesis_info_view gen
+				ON leaves.asset_genesis_id = gen.gen_asset_id
+			WHERE roots.proof_type = 'issuance'
+			GROUP BY gen.asset_id
+		)
+		SELECT asset_id, supply, COUNT(*) as num_leaves
+		FROM asset_supply
+		GROUP BY asset_id, supply`
```

Comment:
why not compile these queries with sqlc?

Reply:
(nit, non blocking)

Reply:
They would then be part of the "production" code, which we don't necessarily want. So I think in tests it's fine to have pure SQL, as long as it's compatible with both SQLite and Postgres.